### PR TITLE
First iteration of async ConnectionManager

### LINKED ISSUE #1068: Connection manager - use async transports
Connection manager should provide for the following:

- Establish peer connection
- Notify once an incoming connection is made
- List connected peers
- Get active connection
- Drop connection
- Shutdown all connections

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

- Changed transport trait to allow concurrent connection upgrades (not
  fully implemented here)
- Added `next` feature and made applicable dependencies optional
- Connection manager responsible for:
  - Starting the Dialer service (Responsible for concurrent dialing
    using the given transport, notifying the connection manager when
    done)
  - Starting the peer listener service (responsible for listening and
    accepting inbound connections over the given transport, and
    notifying the connection manager of successful inbound connections)
  - Creating peer connections (each connected peer has a running service which monitors
    for new multiplexed substreams #l33tnetworkingspeak)
- Added NoiseTransport which can be composed with other transports to
  upgrade connections over noise. Provides a verified peer public key as
  one of it's outputs
- Unit and integration testing (**TODO NB**)

## Motivation and Context

Closes #1068 (New issues #1150, #1151, #1152, #1153)

## How Has This Been Tested?

Tests are TODO - all untested features are currently behind the next feature flag and will not be used by the rest of the codebase until tested. 

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: comms/src/connection_manager/dialer.rs [hunk 1]
```diff
@@ -20,17 +20,378 @@
...
+            request_rx: request_rx.fuse(),
+            cancel_signals: Default::default(),
+            conn_man_notifier,
+            shutdown: Some(shutdown),
+            pending_dial_requests: Default::default(),
+        }
+    }
+
+    pub async fn run(mut self) {
+        let mut pending_dials = FuturesUnordered::new();
```
Comment (philipr-za) [inline comms/src/connection_manager/dialer.rs]: Couldn't this just be part of the Dialer struct so that you don't need to pass it around to the handlers and just call it from `self`?

Comment (sdbondi) [reply comms/src/connection_manager/dialer.rs]: Perhaps - I had troubles with the borrow checker doing it that way in the past, but let me try again

Comment (philipr-za) [reply comms/src/connection_manager/dialer.rs]: I did it in the Transaction Service but might be a subtle difference in the applications

Comment (sdbondi) [reply comms/src/connection_manager/dialer.rs]: So the real reason I didn't do this is because:
1. The box future would need to be `Sync` (because connection manager is spawned) and that's an extra trait bound not needed when it's contained in the method. 
2. the type of future is `BoxFuture<'static, Option<(DialState, DialResult<TTransport::Output>)>>` - this means that the trait bound `TTransport: Transport` needs to be on the Dialer and ConnectionManager struct declarations

These are not insurmountable issues but I think that it ends up being neater to contain it within the method

Comment (philipr-za) [reply comms/src/connection_manager/dialer.rs]: Cool cool
